### ISSUE TITLE
argocd login --sso oauth2: cannot fetch token: 401 Unauthorized

### ISSUE THREAD

**Issue by skan-splunk** [argoproj/argo-cd#12124]
**Describe the bug**

Attempting to use argocd CLI and authenticating with `argocd login <server> --sso` and Okta SSO, and receiving the following output and error in the CLI:  
```
WARN[0005] Failed to invoke grpc call. Use flag --grpc-web in grpc calls. To avoid this warning message, use flag --grpc-web.
Opening browser for authentication
INFO[0007] RequestedClaims: map[groups:essential:true ]
Performing authorization_code flow login: https://{blahhost}&code_challenge_method={blahhash}&redirect_uri=http%3A%2F%2Flocalhost%3A8085%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+groups+offline_access&state={blah}
FATA[0027] oauth2: cannot fetch token: 401 Unauthorized
Response: {"error":"invalid_client","error_description":"Client authentication failed. Either the client or the client credentials are invalid."}
```
After Okta authentication, I get an accepted callback to the URL:`http://localhost:8085/auth/callback?code={blah}&state={blah}` which has the error:
```
oauth2: cannot fetch token: 401 Unauthorized
Response: {&#34;error&#34;:&#34;invalid_client&#34;,&#34;error_description&#34;:&#34;Client authentication failed. Either the client or the client credentials are invalid.&#34;}
```
Authenticating directly into ArgoCD UI works fine, additionally able to authenticate using `--auth-token` flag.
**Expected behavior**

Run `argocd login <server> --sso`, authenticate via okta SSO login, success

**Version**
ArgoCD v2.5.3

**Comment by icecoffee531** [comment]
When I set the oidc provider config to `public:true` it works perfectly (in my case using keycloak)

**Comment by jfdoube** [comment]
Still an issue here as well

**Comment by eranreshef-starkware** [comment]
Ran into this also.
@icecoffee531 I tried your suggestion and it didn't work for me. Did you do anything else I might have missed?

**Comment by billabongrob** [comment]
We've attempted to do the same with Okta and were running into the same issues even with `public: true` 

Setting the cliClientID helped us, FWIW.  
```
    # Some OIDC providers require a separate clientID for different callback URLs.
    # For example, if configuring Argo CD with self-hosted Dex, you will need a separate client ID
    # for the 'localhost' (CLI) client to Dex. This field is optional. If omitted, the CLI will
    # use the same clientID as the Argo CD server
    cliClientID: vvvvwwwwxxxxyyyyzzzz
```
[ref](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/#existing-oidc-provider)

**Comment by stefanandres** [comment]
@billabongrob 

> Setting the cliClientID helped us, FWIW.
What did you do on Okta for this? Did you create another Okta app with localhost as callback url or something? 
It would be great if you could elaborate on what exactly you needed to do to get it working.

**Comment by billabongrob** [comment]
Absolutely, sorry for being vague - end of the day response!  😃 
Yes, we created a Native type OIDC Okta app for the CLI and set it to use Authorization Code + PKCE.  The sign in redirect URI was http://localhost:8085/auth/callback and the sign out redirect URI was http://localhost:8085

**Comment by eranreshef-starkware** [comment]
Thanks @billabongrob your solution worked for me too :tada:

**Comment by zentavr** [comment]
@billabongrob I have Keykloak at my side and adding `Valid redirect URIs`.
- https://argocd-k8s-production.example.io/auth/callback (it was at the beginning)
- http://localhost/auth/callback (added)
- http://localhost:8085/auth/callback (added)

Also I'd adjusted `Web origins`:
- https://argocd-k8s-production.example.io (it was at the beginning)
- http://localhost (added)
- http://localhost:8085 (added)

Other settings are the next:
![Знімок екрана 2024-07-03 о 04 05 44](https://github.com/argoproj/argo-cd/assets/595334/f6a93ac0-2115-408a-bd99-c66da023640d)

There is still the issue:
```
$ argocd login argocd-k8s-production.wifimap.io --sso --name argocd-k8s-production
WARN[0003] Failed to invoke grpc call. Use flag --grpc-web in grpc calls. To avoid this warning message, use flag --grpc-web.
Opening browser for authentication
Performing authorization_code flow login: https://sso.example.io/realms/Example/protocol/openid-connect/auth?access_type=offline&client_id=argocd-production-k8s-shared&code_challenge=iI8mfDBZNByRXPnZUruwZhJ5oLHnZouqrumI1KWSjYs&code_challenge_method=S256&redirect_uri=http%3A%2F%2Flocalhost%3A8085%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+groups+offline_access&state=vHlgeXwhAfriUvkGtFrWCPVK
FATA[0013] oauth2: "unauthorized_client" "Invalid client or Invalid client credentials"
```

Have that after entering login+password in the pop up browser window.

**Comment by zentavr** [comment]
So the reply to my question above:
https://github.com/argoproj/argo-cd/issues/10029#issuecomment-1673301634

For keykloak this needs to be done:

![Знімок екрана 2024-07-03 о 04 14 13](https://github.com/argoproj/argo-cd/assets/595334/b0da1cbe-f417-48a7-956e-52339c598201)

Client authentication: `OFF`

**Comment by dmakeienko** [comment]
I have a similar problem and despite configuring Keyclock as per  @zentavr suggestion, cli login not working, but rather send a following error:
`invalid_scope: Invalid scopes: openid profile email groups offline_access`.
The strange part is that I didn't add `offline_access` to scope

**Comment by albundy83** [comment]
I have tried but I have an issue and I see the error  `invalid_request: Missing parameter: code_challenge_method` after 5 minutes.

**Comment by manueljishi** [comment]
@albundy83 I'm experiencing the same issue with keycloak and argocd. Did you manage to solve your issue? Whenever the token keycloak generates expires, I get that error instead of a refresh token

**Comment by albundy83** [comment]
Hello, have you tried the updated documentation from [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/keycloak/#keycloak-and-argocd-with-pkce).
Especially, did you disable the Client authentication in keycloak?
Be sure to correctly set all parameters

**Comment by manueljishi** [comment]
Hey, I had followed the documentation but I couldn't make it work but also your comment gave me the idea to check the argocd version. I was running version 2.11 which apparently didn't support PKCE so I wanted to leave a comment here in case anyone has the same issue in the future.
Many thanks!

**Comment by dmakeienko** [comment]
> Hello, have you tried the updated documentation from [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/keycloak/#keycloak-and-argocd-with-pkce). Especially, did you disable the Client authentication in keycloak? Be sure to correctly set all parameters

You are correct, ArgoCD v2.5.3 didn't have that detailed documentation at that time issue occured. 
Also, it's fixed with newer version of ArgoCD  v2.13+ with PKCE flow.
